docs/SPEC.md

```diff
@@ -729,29 +729,45 @@ It is not possible to access an object's property using an arbitrary expression.
     DotExpression           = "." identifer
     MemberBracketExpression = "[" string_lit "]" .
 
+#### Conditional Expressions
+
+Conditional expressions evaluate a boolean-valued condition and if the result it _true_,
+the expression following the `then` keyword is evaluated and returned.
+Otherwise the condition evaluates to _null_ or _false_,
```

Comment:
Not sure why the condition should ever evaluate to _null_. What's the use case?

Reply:
In general I think null values can appear almost anywhere that a value appears. In the case of `map`, for example:
```
   ... |> map(fn: (r) => {code: if r.code == 0 then "green" else "red"})
```

Reply:
I think as long as we assert that if statements have to type check to a boolean this is fine, but every value is nullable in flux at the moment so the boolean could be null.

If we want to take the type system further, we can differentiate between nullable columns and non-nullable ones. Anything coming from the storage engine is non-nullable, but some methods like join or pivot will add a nullable property onto everything and it might be too "pure" for many users.

Reply:
Curious to hear about this one too. Otherwise this LGTM

Reply:
Ok, I'll remove the `or null` bit.

So I guess `exists code r` would evaluate to true if either the column `code` exists and is null, or if the column `code` does not exist?

Will we be doing any type checking at runtime once the schema of tables is known, or is type checking only a compile-time thing?

This whole discussion also seems relevant to the stream index functions, since we'll want to figure out how to avoid nulls leaking out there.

Reply:
> So I guess exists code r would evaluate to true if either the column code exists and is null, or if the column code does not exist?

Yes, the semantics of null are the value is missing, whether all rows or just some of the rows are missing the value is not an important distinction.

> Will we be doing any type checking at runtime once the schema of tables is known, or is type checking only a compile-time thing?

Type checking will happen at compile time and runtime, this is what I mean when I say we want Flux to be gradually typed. If the compiler can do the type checking then it will insert the runtime type checks instead.

> This whole discussion also seems relevant to the stream index functions, since we'll want to figure out how to avoid nulls leaking out there.

I agree, the `record` function needs to not create properties on the returned object if it is null.

Reply:
Ok, I think I get it.

We might have an expression like this, say as the lambda in a `map`:
```
if exists _value r then r._value else 0
```
At compile time, we don't know what the fields of `r` are, so we instrument the output of the compiler to be something like:
```
if 
    exists _value r 
then 
    assert has_field _value r
    typeassert r._value int
    r._value
else
    0
```
So at compile time, the type checking is a pass over the IR.  At runtime, type checks are evaluated as part of the program.  Is this close to how you're thinking it will work?

Reply:
What is the difference between `assert has_field` and `exist _value r` checks?

I was thinking something slightly different:

Case `r`'s type is unknown:

```
if
    exists _value r 
then 
    typeassert r._value int
    r._value
else
    0
```


Case `r`'s type is known to have the `_value` property with type `int`:

```
r._value
```

Reply:
> What is the difference between assert has_field and exist _value r checks?

Yes, you're right, that first assert is unnecessary.  I think we're on the same page.

Reply:
@nathanielc 

So I guess for filters a common idiom would be something like:
```
filter(fn: (r) => if exists id r and r.id == 100 then true else false)
```
I wonder if we should provide some syntactic sugar for this common case?

I think I like this approach for InfluxDB sources, but what about SQL?  I have a couple concerns.

Suppose a user writes a Flux script that accesses SQL data, and maybe joins it to data in their InfluxDB source.  Their test data has no nulls, and the script works fine.  In production (potentially a long while later) nulls appear, and the script now breaks.  This could be painful for users.

My second concern regards how we optimize Flux queries for SQL sources.  I'd imagine we want to allow the user to write Flux in a natural way and let the planner push things down into the SQL source.  How should we translate something like `filter(fn: (r) => r.id == 100)` into SQL?  If `r.id` being null is a type error, then I guess we'd want SQL to throw an exception, and I'm not sure how to do that.  With syntactic sugar for nulls+predicates, maybe it gets easier.